backend/app.py

```diff
@@ -171,49 +170,167 @@ def jqQuery(objtype: str, query: str, objlist=None, ) -> Union[Dict, List]:
+    concept_set_ids = codeset_id.split('|')
+    concept_set_ids = [int(x) for x in concept_set_ids]
+    df_concept_set_members = DS['concept_set_members']
+    df_concept_ancestor = DS['concept_ancestor']
+    df_concept_set_members_i = df_concept_set_members[df_concept_set_members['codeset_id'].isin(concept_set_ids)]
+    df_concept_ancestor_i = df_concept_ancestor[
+        (df_concept_ancestor.ancestor_concept_id.isin(df_concept_set_members_i.concept_id)) &
+        (df_concept_ancestor.descendant_concept_id.isin(df_concept_set_members_i.concept_id))]
+
+    # Transform: get essential information
+    # todo: account for max_levels_of_separation, or just get direct parent/child relationships
+    req_concepts: Dict = concepts_and_concept_sets['concepts']
+    req_csets: Dict = concepts_and_concept_sets['concept_sets']
+    table_data = []
+    for concept_id, concept in req_concepts.items():
+        if concept_id not in table_data:
+            df_concept_ancestor_i2 = df_concept_ancestor_i[
+                df_concept_ancestor_i['descendant_concept_id'] == int(concept_id)]
+            df_concept_ancestor_i2 = df_concept_ancestor_i2.sort_values(['min_levels_of_separation'], ascending=True)
+            ancestors_d_list: List[Dict] = df_concept_ancestor_i2.to_dict(orient='records')
+            for d in ancestors_d_list:
+                ancestor = str(d['ancestor_concept_id'])
+                new_d = {
+                    'ConceptID': concept['concept_id'],
+                    'AncestorID': ancestor,
+                    'min_levels_of_separation': d['min_levels_of_separation'],
+                }
+                for concept_set_id in [str(x) for x in concept_set_ids]:
+                    cset_concepts: List[str] = [str(x) for x in req_csets[int(concept_set_id)]['concepts'].keys()]
+                    new_d[concept_set_id] = \
+                        'O' if ancestor in [str(x) for x in cset_concepts] else 'X'
+                table_data.append(new_d)
+
+    # Transform: Visualize indent
+    table_data_2 = []
+    # todo: temp initializing these vars here until later refactor
+    concept_ids = []
+    cset_ids = []
+    for row in table_data:
+        row['ConceptID'] = row['AncestorID']
+        row['ConceptID'] = f"{indent_string * row['min_levels_of_separation']}{row['ConceptID']}"
+        del row['AncestorID']
+        del row['min_levels_of_separation']
+        table_data_2.append(row)
+        # todo: temp until later refactor
+        concept_ids.append(row['ConceptID'])
+        for cset_id in [x for x in row.keys() if x != 'ConceptID']:
+            cset_ids.append(cset_id)
+
+    # get labels
+    # todo: this whole route needs to be refactored eventually.
+    # todo: we may want to pickle some lookup maps maybe
+    concept_ids_set = set([int(x.replace('-', '')) for x in concept_ids])
+    concept_df = CONCEPT[CONCEPT['concept_id'].isin(concept_ids_set)]
```

Comment:
@Sigfried This was surprisingly fast. I basically look at all the codesets and concepts that are relevant to the request, and then I filter the appropriate tables, and then map the ID to the label.

Eventually though this will be refactored I think and the labels will simply be there and not need to be mapped. If we do need ID->label maps, we can also create that map, pickle, and cache them too